### PR TITLE
Eager pull fan-in stages.

### PR DESCRIPTION
**Pull request by ktoso**
In order to bring in https://github.com/akka/akka/pull/19198

### CODE REVIEW

**Comment by ktoso** [comment]
Got reproducable failures locally... don't want to include it in 2.0.1

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template [hunk 1]
```diff
@@ -56,17 +65,13 @@ class ZipWith1[[#A1#], O] (zipper: ([#A1#]) ⇒ O) extends GraphStage[FanInShape
 
     setHandler(out, new OutHandler {
       override def onPull(): Unit = {
-        pending = shape.inlets.size
-        if (willShutDown) completeStage()
-        else {
-            [#pull(in0)#
-            ]
-        }
+        pending += shape.inlets.size
```
Comment (viktorklang) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template]: `inlets` is an immutable.Seq so `size` could most definitely be `O(n)` so perhaps not do that for every one of these calls?

Comment (viktorklang) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template]: But looking at the FanInShape *impl* it does look like a Vector :)

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template [hunk 2]
```diff
@@ -56,17 +65,13 @@ class ZipWith1[[#A1#], O] (zipper: ([#A1#]) ⇒ O) extends GraphStage[FanInShape
...
+        if (pending == ##0) pushAll()
       }
     })
   }
 
-  override def toString = "Zip"
+  override def toString = "ZipWith1"
```
Comment (viktorklang) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template]: ZipWith##0 ?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 3]
```diff
@@ -643,7 +640,16 @@ object Concat {
   /**
    * Create a new `Concat`.
    */
-  def apply[T](inputPorts: Int = 2): Concat[T] = new Concat(inputPorts)
+  def apply[T](inputCount: Int = 2) = GraphDSL.create() { implicit builder ⇒
+    import GraphDSL.Implicits._
+    val concat = builder.add(new Concat[T](inputCount))
+    val ds = concat.inSeq.map { inlet ⇒
+      val detacher = builder.add(new Detacher[T])
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: Isn't it possible to cache the Detacher instance and add it for every element?

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template (hunk 2 above)
Comment (ktoso) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template]: Why? We want that number to be ZipWith2, ZipWith3, ZipWith4 etc.
`##0` hardcodes it to 0.

**Comment by ktoso** [comment]
@2m is reviving his PR back in https://github.com/akka/akka/pull/19198 now,
closing this one.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala (hunk 3 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: This is already done, it is called `def detacher[T]` now, which should be used here.

**Pull request closed without merging**